### PR TITLE
Allow user to set Language Hints for Note Editor fields

### LINKED ISSUE #3496: Automatically change a keyboard layout on focus
Originally reported on Google Code with ID 2600

```
It would be awesome to make the app remember the last keyboard layout for each field
when you create new cards. E.g. you have Russian and English translation fields. So,
the layout (in the apps SwiftKey and Swype as most popular, for example) switches to
Russian in the first field and bakc to English in the second automatically when the
fields get focus.
```

Reported by `pyastolov` on 2015-05-08 08:04:17

### PR DESCRIPTION
**Pull request by david-allison-1**
## Draft

I've removed the draft flag - I'd like to get this in and refactor if it gets traction and then refactor (if feasible)

----

This needs investigation as to whether this can be simplified via 

```java
InputMethodSubtype ims = InputMethodManager.getCurrentInputMethodSubtype().getLocale();
```

If this works (50/50 - Android IME options are flaky), then the field editor fragment and locale selection screen can be removed and this feature is much smaller.

## To Review

* This does not handle an "unset" - I don't feel this is necessary, but that's borderline.
* We use `Locale.getAvailableLocales()` for the locale list, we currently use the system language. Do we want to use the AnkiDroid language here?
* Does this need to be split up further to make it reviewable? Where are the pain points currently?

## Purpose / Description
Feature request: allow the keyboard to remember the last input language on a field.

Sadly, this didn't seem possible, but it is possible to select and keep a "language hint"

## Fixes
Fixes #3496 

## Approach
Add a field into the "field" JSON specifying the language. Add a context menu on the note editor, and menu in the fields screen to set

## How Has This Been Tested?

Manually tested - including with Don't Keep Activities.

This works after syncing.

This does not work with SwiftKey, which is disappointing.

Screenshots and video incoming

## Learning

Initial StackOverflow research said this wasn't possible, they were mistaken

https://developer.android.com/reference/android/widget/TextView#setImeHintLocales(android.os.LocaleList)

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by david-allison-1** [comment]
Video: https://drive.google.com/file/d/19NNIilMRJbNmduQCjAYy1Lk7wZtCRbt0/view

image

image

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java [hunk 1]
```diff
@@ -537,6 +545,54 @@ public void onBackPressed() {
             case ModelEditorContextMenu.FIELD_TOGGLE_STICKY:
                 toggleStickyField();
                 break;
+            default: {
+                //need this as we can't switch on a @RequiresApi
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java:549]: Flagging in case there's a better way here

**Comment by mikehardy** [comment]
Quite the pile of conflicts can review when clean :)

**Comment by david-allison-1** [comment]
Rebased - that was unpleasant (I'd done pretty much the same refactoring twice for different reasons)

**Comment by david-allison-1** [comment]
@mikehardy Waiting for CI

**Comment by mikehardy** [comment]
Bam! New 2.13 feature

**Pull request merged**